### PR TITLE
Packages: Add more npm release types to the GitHub workflow

### PR DESCRIPTION
**Pull request by gziolo**
## What?

Follow-up for #40976.

This PR adds more release types to the existing GitHub Workflow that handles publishing npm packages:
- development (`next` dist tag)
- bugfix (`latest` dist tag)

Together with the preexisting `wp` (`wp/X.Y` dist tag) release type it covers all supported manual releases. 

See [Packages Releases to npm and WordPress Core Updates](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/release.md#packages-releases-to-npm-and-wordpress-core-updates) to learn more about release types.

Note: The last remaining release type [Synchronizing Gutenberg Plugin](https://github.com/WordPress/gutenberg/compare/update/npm-publish-workflow-types?expand=1#synchronizing-gutenberg-plugin) is handled separately in a different workflow.

## Why?

It will simplify the process of publishing:
- bugfixes for the existing npm packages unrelated to the Gutenberg plugin code
- development releases whenever we want to test how new changes work with the npm infrastructure

## How?

It's nearly identical to https://github.com/WordPress/gutenberg/pull/40976 and how `wp` release type is handled. The only difference is that new release types use a different script at the end of the process according to the existing documentation:

- [Development releases](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/release.md#development-releases)
- [Bugfix releases](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/release.md#standalone-bugfix-package-releases)

## Testing Instructions

It's difficult to test in the development branch in the Gutenberg repository because the UI won't update until we merge this PR...

I tested in the forked version of this repository with a simplified script that doesn't use secret tokens and doesn't run the publishing step.

The flow is nearly identical to what I documented for the preexisting `wp` release type:

https://github.com/WordPress/gutenberg/pull/40976#issuecomment-1125108897

## Screenshots or screencast 

The updated `Run workflow` form:

Screenshot 2022-05-12 at 18 37 04

The list presenting workflows triggered manually. You can also check at https://github.com/gziolo/gutenberg/actions/workflows/publish-npm-packages.yml.

Screenshot 2022-05-12 at 18 45 56

Example of the successful workflow in the test repo with the release type `development`:

Screenshot 2022-05-12 at 18 46 16

Example of the successful workflow in the test repo with the release type `bugfix`:

Screenshot 2022-05-12 at 18 46 27

### CODE REVIEW

**Comment by gziolo** [comment]
@ockham or @desrosj, is there any simple way to avoid some code duplication between jobs in the GitHub workflow? They differ only in the last step for now. I don't mind leaving it as is, but it made me curious to hear an opinion from the experts 😄

**Comment by gziolo** [comment]
The remaining task once this PR lands:

- run and document how the new workflow works for manual publishing
- ensure that the release type `wp` works correctly with older `wp/5.x` branches that use Node 10 or 12
- update documentation for Block Editor Release Process for Major Releases - at least the [Managing Weekly Beta and RC releases](https://make.wordpress.org/core/handbook/about/release-cycle/block-editor-release-process-for-major-releases/#managing-weekly-beta-and-rc-releases)
- post at https://make.wordpress.org/core/ with the summary of the changes applied to npm publishing
  - new npm user for publishing https://www.npmjs.com/~gutenbergplugin and approval from the Gutenberg core team (should we add all core committers or a subset of them?) 
  - automatic npm publishing tied to Gutenberg releases
  - manually triggered npm publishing for WP major releases and the trick with `dist-tag` to help with the flow

**Comment by ockham** [comment]
> @ockham or @desrosj, is there any simple way to avoid some code duplication between jobs in the GitHub workflow? They differ only in the last step for now. I don't mind leaving it as is, but it made me curious to hear an opinion from the experts 😄

I think it should be possible to:
- unify the jobs into one (that's just named `Release`, or maybe `Release (${{ github.event.inputs.release_type }})` works)
- keep all three versions of the final (publishing) step
- remove the `if: ${{ github.event.inputs.release_type == ... }}` clauses that checks which release type we're dealing with from the job level, and move them to those final steps.

IIRC, I did something similar in other workflows 🤔

**Comment by ockham** [comment]
Alternatively, what might work is to three separate jobs that each just contain the publishing step, and move the other steps to a `setup` job (or whatever you wanna call it), and have each of the three different publishing jobs depend on it via `needs`.

Edit: Since you need files from the setup job in the publish jobs, I think you would need to "transfer" them using [`upload-artifact`](https://github.com/actions/upload-artifact) and [`download-artifact`](https://github.com/actions/download-artifact) actions.

**Comment by walbo** [comment]
> @ockham or @desrosj, is there any simple way to avoid some code duplication between jobs in the GitHub workflow? They differ only in the last step for now. I don't mind leaving it as is, but it made me curious to hear an opinion from the experts 😄

@gziolo Have you look into reusable workflows? Seems to me like a good usecase. https://docs.github.com/en/actions/using-workflows/reusing-workflows

**Comment by gziolo** [comment]
> I think it should be possible to:
> 
> * unify the jobs into one (that's just named `Release`, or maybe `Release (${{ github.event.inputs.release_type }})` works)
> * keep all three versions of the final (publishing) step
> * remove the `if: ${{ github.event.inputs.release_type == ... }}` clauses that checks which release type we're dealing with from the job level, and move them to those final steps.
> 
> IIRC, I did something similar in other workflows 🤔

That'd work. Why didn't I think about it in the first place? 😄 Thank you so much @ockham. I checked other workflows and it should be exactly what I need.

> Have you look into reusable workflows? Seems to me like a good usecase.
> https://docs.github.com/en/actions/using-workflows/reusing-workflows

@walbo, that's an interesting option. It's more advanced and I guess it would suffer from the challenge described by @ockham:

> Since you need files from the setup job in the publish jobs, I think you would need to "transfer" them using [upload-artifact](https://github.com/actions/upload-artifact) and [download-artifact](https://github.com/actions/download-artifact) actions.

**Comment by gziolo** [comment]
Everything works as expected after all the tweaks were applied. I tested with the forked repository (https://github.com/gziolo/gutenberg/actions/workflows/publish-npm-packages.yml) and the last step was simplified to avoid npm publishing.

**Review by adamziel — APPROVED** [review]
The log, the reasoning, and the diff all look good – thank you for working on that @gziolo 🎉

**Comment by gziolo** [comment]
I documented next steps at https://github.com/WordPress/gutenberg/discussions/37820#discussioncomment-2759429.

Let’s test the `wp` release type tomorrow 😃

**Pull request merged**
